streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimestampedKeyValueStore.java

```diff
@@ -0,0 +1,212 @@
+        return this.map.get(key);
+    }
+
+    @Override
+    public synchronized void put(final K key,
+                                 final ValueAndTimestamp<V> valueAndTimestamp) {
+        if (valueAndTimestamp == null || valueAndTimestamp.value() == null) {
+            this.map.remove(key);
+        } else {
+            this.map.put(key, valueAndTimestamp);
+        }
+    }
+
+    @Override
+    public synchronized void put(final K key,
+                                 final V value,
+                                 final long timestamp) {
+        if (value == null) {
+            this.map.remove(key);
+        } else {
+            this.map.put(key, new ValueAndTimestampImpl<>(value, timestamp));
+        }
+    }
+
+    @Override
+    public synchronized ValueAndTimestamp<V> putIfAbsent(final K key,
+                                                         final ValueAndTimestamp<V> valueAndTimestamp) {
+        final ValueAndTimestamp<V> originalValueAndTimestamp = get(key);
+        if (originalValueAndTimestamp == null) {
+            put(key, valueAndTimestamp);
+        }
+        return originalValueAndTimestamp;
+    }
+
+    @Override
+    public synchronized ValueAndTimestamp<V> putIfAbsent(final K key,
+                                                         final V value,
+                                                         final long timestamp) {
+        return putIfAbsent(key, new ValueAndTimestampImpl<>(value, timestamp));
+    }
+
+    @Override
+    public synchronized void putAll(final List<KeyValue<K, ValueAndTimestamp<V>>> entries) {
+        for (final KeyValue<K, ValueAndTimestamp<V>> entry : entries) {
+            put(entry.key, entry.value);
+        }
+    }
+
+    @Override
+    public synchronized ValueAndTimestamp<V> delete(final K key) {
+        return this.map.remove(key);
+    }
+
+    @Override
+    public synchronized KeyValueIterator<K, ValueAndTimestamp<V>> range(final K from,
+                                                                        final K to) {
+        return new DelegatingPeekingKeyValueIterator<>(
+            name,
+            new InMemoryKeyValueIterator<>(this.map.subMap(from, true, to, true).entrySet().iterator()));
+    }
+
+    @Override
+    public synchronized KeyValueIterator<K, ValueAndTimestamp<V>> all() {
+        final TreeMap<K, ValueAndTimestamp<V>> copy = new TreeMap<>(this.map);
+        return new DelegatingPeekingKeyValueIterator<>(name, new InMemoryKeyValueIterator<>(copy.entrySet().iterator()));
+    }
+
+    @Override
+    public long approximateNumEntries() {
+        return this.map.size();
+    }
+
+    @Override
+    public void flush() {
+        // do-nothing since it is in-memory
+    }
+
+    @Override
+    public void close() {
+        this.map.clear();
+        this.open = false;
+    }
+
+    static class InMemoryKeyValueIterator<K, V> implements KeyValueIterator<K, V> {
```

Comment:
Ah. Actually this code should not be here -- it's the same as `InMemroyKeyValueStore.InMemoryKeyValueIterator` -- that's why `private` modifier was removed.

Will remove this nested class to avoid code duplication.